### PR TITLE
【Hackathon No.29】add RFC for PixelUnshuffle

### PR DESCRIPTION
**Pull request by BrilliantYuKaimin**
增加paddle.nn.PixelUnshuffle设计文档。

### CODE REVIEW

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+        self.factor = downscale_factor
+        self.data_format = data_format
+
+    def forward(self, x):
+        n, c, h, w = x.shape
+        x = paddle.reshape(x, [n, c, h / self.factor, self.factor, w / self.factor, self.factor])
+        x = paddle.transpose(x, [0, 1, 3, 5, 2, 4])
+        x = paddle.reshape(x, [n, c * self.factor * self.factor, h / self.factor, w / self.factor])
+        return x
+```
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:122]: 这部分可以放在飞桨现状

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+
+参考`paddle.nn.PixelShuffle`来实现`paddle.nn.PixelUnshuffle`，顺便实现`paddle.nn.functional.pixel_unshuffle`。
+
+# 六、测试和验收的考量
+
+考虑测试的情况：
+- 与PyTorch的结果的一致性；
+- 反向传播的正确性；
+- 错误检查：`downscale_factor`不是正整数或不同时整除宽度和高度时能正确抛出异常。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:149]: 这部分需要修改，请参考这个PR中描述：https://github.com/PaddlePaddle/community/pull/40

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+- **downscale_factor** (int) - 空间分辨率缩小的比例，需要同时整除宽度和高度；
+- **data_format**(str) - 指定输入的数据格式，输出的数据格式将与输入保持一致，可以是"NCHW"和"NHWC"。N是批尺寸，C是通道数，H是特征高度，W是特征宽度。默认值："NCHW"。
+
+## 底层OP设计
+PixelUnshuffle与飞桨中已有的PixelShuffle操作在逻辑上是一样的，故可参考PixelShuffle的OP设计来实现PixelUnshuffle的底层OP。
+
+## API实现方案
+
+参考`paddle.nn.PixelShuffle`来实现`paddle.nn.PixelUnshuffle`，顺便实现`paddle.nn.functional.pixel_unshuffle`。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:142]: 这部分需要修改，请参考这个PR中描述：https://github.com/PaddlePaddle/community/pull/40

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        n, c, h, w = x.shape
+        x = paddle.reshape(x, [n, c, h / self.factor, self.factor, w / self.factor, self.factor])
+        x = paddle.transpose(x, [0, 1, 3, 5, 2, 4])
+        x = paddle.reshape(x, [n, c * self.factor * self.factor, h / self.factor, w / self.factor])
+        return x
+```
+
+# 四、对比分析
+
+无论是C++实现还是组合API实现，其逻辑都是十分简单的，故考虑使用C++编写新的算子以期取得更高的效率。
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:126]: 这部分需要修改，请参考这个PR中描述：https://github.com/PaddlePaddle/community/pull/40

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+
+```python
+def shuffle_unit(self, x, downscale_factor):
+    with tf.variable_scope('pixel_unshuffle_unit'):
+        n, h, w, c = x.get_shape().as_list()
+        x = tf.reshape(x, shape=tf.convert_to_tensor([tf.shape(x)[0], h / downscale_factor, downscale_factor, w / downscale_factor, downscale_factor, c]))
+        x = tf.transpose(x, tf.convert_to_tensor([0, 1, 3, 5, 2, 4]))
+        x = tf.reshape(x, shape=tf.convert_to_tensor([tf.shape(x)[0], h / downscale_factor, w / downscale_factor, c * downscale_factor * downscale_factor]))
+
+```
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:103]: 这部分需要增加内容，请参考这个PR中描述：https://github.com/PaddlePaddle/community/pull/40

**Review by shiyutang — CHANGES_REQUESTED** [review]
方案整体完备，但是实现细节等需要步骤化说明，请修改方案后再提交代码。修改示例参照：https://github.com/PaddlePaddle/community/blob/master/rfcs/APIs/20200301_api_design_for_quantile.md

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 1 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md]: > 这部分可以放在飞桨现状

完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 2 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md]: > 这部分需要修改，请参考这个PR中描述：#40

完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 3 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:175]: > 这部分需要修改，请参考这个PR中描述：#40

完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 4 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md]: > 这部分需要修改，请参考这个PR中描述：#40

完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 5 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md]: > 这部分需要增加内容，请参考这个PR中描述：#40

完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+                                         -4 /* oh */, -2 /* ow */});
+  const auto input_permuted = input_reshaped.permute(permutation);
+
+  std::vector<int64_t> final_shape(self.sizes().begin(), self_sizes_batch_end);
+  final_shape.insert(final_shape.end(), {oc, oh, ow});
+  return input_permuted.reshape(final_shape);
+}
+
+}} // namespace at::native
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:102]: 需要进一步增加对torch方案的步骤化描述

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+        pass
+```
+
+# 六、测试和验收的考量
+
+- 是否同时支持静态图和动态图；
+- 是否同时支持CPU和GPU平台；
+- 测试不同张量类型下的表现；
+- 对全部入参进行参数有效性和边界值测试，确定每个入参都可以正确生效；
+- 前向计算的正确性（与组合API实现比较、与PyTorch比较）；
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:213]: 本地torch，CI中使用numpy实现的算法进行对比

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+- 反向计算的正确性；
+- 当传入的`downscale_factor`不合法（不是正整数、不同时整除高度和宽度）时会抛出异常并有友好的提示。
+
```
Comment (shiyutang) [inline rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:216]: 错误检查：输入维度，downscale_factor > 0等

**Review by shiyutang — CHANGES_REQUESTED** [review]
参考https://github.com/PaddlePaddle/community/pull/40 进行同步修改

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 6 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md:102]: 完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 7 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md]: 完成

File: rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md (hunk 8 above)
Comment (BrilliantYuKaimin) [reply rfcs/APIs/20220317_api_design_for_PxielUnshuffle.md]: 完成

**Review by shiyutang — APPROVED** [review]
LGTM

**Comment by shiyutang** [comment]
实现内容可以存放在paddle/phi/kernels/cpu(gpu)/ 下，注册则在fluid/oprerator下。如下图所示：
以TraceOp为例：
c++ op开发方式可以参见开发指南：https://www.paddlepaddle.org.cn/documentation/docs/zh/develop/dev_guides/api_contributing_guides/new_cpp_op_cn.html

**Comment by shiyutang** [comment]
请对文档中的实现内容位置进行删除或者对应修改

**Comment by BrilliantYuKaimin** [comment]
> 请对文档中的实现内容位置进行删除或者对应修改

因为cpu和gpu的kernel是一样的，所以参考pixel_shuffle把实现放在`paddle/phi/kernels/impl/pixel_unshuffle_kernel_impl.h`中。

**Comment by shiyutang** [comment]
如果实现和设备无关，应当放在kernel根目录下，如下所示：
image

image

**Comment by shiyutang** [comment]
请提交一个新的PR对相应内容进行修改
